File: sucraft-core/src/main/java/org/sucraft/core/common/plugin/SuCraftPlugin.java
package org.sucraft.core.common.plugin;

import java.util.Objects;

import org.bukkit.Bukkit;
import org.bukkit.NamespacedKey;
import org.bukkit.event.Listener;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.sucraft.core.common.log.DefaultLogTexts;
import org.sucraft.core.common.log.SuCraftLogger;

public abstract class SuCraftPlugin extends JavaPlugin {
	
	// Enable and disable
	
	@Override
	public void onEnable() {
		suCraftLogger = new SuCraftLogger(this);
		suCraftLogger.info(DefaultLogTexts.pluginEnabling);
		onPluginPreEnable();
		onPluginEnable();
		suCraftLogger.info(DefaultLogTexts.pluginEnabled);
	}
	
	@Override
	public void onDisable() {
		suCraftLogger.info(DefaultLogTexts.pluginDisabling);
		onPluginDisable();
		suCraftLogger.info(DefaultLogTexts.pluginDisabled);
	}
	
	/**
	 * The logger of this plugin has been created, nothing else has been done
	 * The plugin singleton instance should be initialized in this method
	 */
	protected abstract void onPluginPreEnable();
	
	protected void onPluginEnable() {}
	
	protected void onPluginDisable() {}
	
	// Logger
	
	private @Nullable SuCraftLogger suCraftLogger;
	
	public @NotNull SuCraftLogger getSuCraftLogger() {
		return Objects.requireNonNull(suCraftLogger, "The logger has not been initialized yet");
	}
	
	// Convenience methods for doing plugin-based things
	
	public void registerEvents(@NotNull Listener listener) {
		Bukkit.getPluginManager().registerEvents(listener, this);
	}
	
	public @NotNull NamespacedKey getNamespacedKey(@NotNull String key) {
		return new NamespacedKey(this, key);
	}
	
}

File: sucraft-core/src/main/kotlin/org/sucraft/core/common/sucraft/plugin/SuCraftPlugin.kt
package org.sucraft.core.common.sucraft.plugin

import org.bukkit.plugin.java.JavaPlugin
import org.sucraft.core.common.bukkit.log.NestedLogger
import org.sucraft.core.common.sucraft.log.SuCraftLogTexts
import org.bukkit.Bukkit
import org.bukkit.NamespacedKey
import org.bukkit.event.Listener
import org.sucraft.core.common.general.pattern.SingletonContainer

abstract class SuCraftPlugin : JavaPlugin() {

	// Enable and disable

	override fun onEnable() {

		// Mark as enabling
		enabling = true

		// Set the singleton instance
		@Suppress("UNCHECKED_CAST")
		(this::class.java.declaredClasses.first { it.simpleName.equals("Companion") }!!.kotlin.objectInstance as SingletonContainer<out SuCraftPlugin>).setInstanceUntyped(this)
		// Create the logger
		nestedLogger = NestedLogger.create(this)

		// Print to console and call enable subroutine
		nestedLogger!!.info(SuCraftLogTexts.pluginEnabling)
		onSuCraftPluginEnable()
		nestedLogger!!.info(SuCraftLogTexts.pluginEnabled)

		// Mark as finished enabling
		enabling = false

	}

	override fun onDisable() {

		// Mark as disabling
		disabling = true

		// Print to console and call disable subroutine
		nestedLogger!!.info(SuCraftLogTexts.pluginDisabling)
		onSuCraftPluginDisable()
		nestedLogger!!.info(SuCraftLogTexts.pluginDisabled)

		// Mark as finished disabling
		disabling = false

	}

	protected open fun onSuCraftPluginEnable() {}
	protected open fun onSuCraftPluginDisable() {}

	private var enabling = false
	private var disabling = false

	// Logger

	private var nestedLogger: NestedLogger? = null
	fun getNestedLogger() = nestedLogger ?: throw IllegalStateException("The nested logger has not been initialized yet")

	// Convenience methods for doing plugin-based things

	fun registerEvents(listener: Listener) {
		Bukkit.getPluginManager().registerEvents(listener, this)
	}

	fun getNamespacedKey(key: String): NamespacedKey {
		return NamespacedKey(this, key)
	}

}